enhancements/monitoring/alerting-consistency.md

```diff
@@ -0,0 +1,195 @@
+
+Alerts are a mess.  There's not really formal guidelines to follow when
+implementing alerts for a given component.  These are the guidelines.
+
+## Motivation
+
+Improve Signal to noise of alerts.  Align alert levels "Critical, Warning, Info"
+with operational expectations to allow easier triaging.  Ensure current and
+future alerts have documentation covering problem/investigation/solution steps.
+
+### Goals
+
+* Define clear criteria for designating an alert 'critical'
+* Define clear criteria for designating an alert 'warning'
+* Define minimum time thresholds before prior to triggering an alert
+* Establish clear documentation guidelines
+* Outline operational triage of firing alerts
+
+### Non-Goals
+
+* Define specific alert needs
+* Implement user-defined alerts
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an OpenShift developer, I need to ensure that my alerts are appropriately
+tuned to allow end-user operational efficiency.
+
+#### Story 2
+
+As an OpenShift administrator, I need to have confidence that alerts are
+actually informative and actionalble to maintain my clusters.
+
+### Implementation Details/Notes/Constraints [optional]
+
+This enhancement will require some developers and administrators to rethink
+their concept of alerts and alert levels.  Many people seem to have an
+intuition about how alerts should work that doesn't necessarily match how
+alerts actually work or how others believe alerts should work.
+
+Formalizing this guidance into a type of template or specification will allow
+developers and administrators to speak a common language and have a common
+understanding, even if the concepts seem initially unituitive.
+
+### Risks and Mitigations
+
+People will make wrong assumptions about alerts and how to handle them.  This
+is not unique to alerts, people often believe certain components should behave
+in a particular way in which they do not.  This is an artifact of poor
+documentation that leaves too much to the imagination.
+
+We can work around this problem with clear and concise documentation.
+
+## Design Details
+
+### Critical Alerts
+
+TL/DR:  Fix this now or cluster dies.
+
+Timeline:  ~5 minutes.
+
+Critical Alerts should be reserved for situations that jeopardize the health
+if not immediately addressed or if the cluster is already in a critical state.
+
+Some critical states are:
+* loss or impending loss of etcd-quorum.
+* inability to route application traffic externally or internally
+(dataplane disruption)
+* inability to start any new pod other than capacity.  EG, if the API server
+were to restart or need to be rescheduled, it would not able to start on another
+host
+
+In otherwords, critical alerts are something that require someone to get out
+of bed in the middle of the night and fix something **right now** or they will
+be faced with a disaster.
+
+An example of something that is NOT a critical alert:
+* MCO and/or related components are completely dead/crash looping.
```

Comment:
That's to say that MCO itself is not relevant to the clusters health, which is not true for all cluster setups

Reply:
The MCO being down does not create a disaster situation.  The MCO does not need to be alive to run existing workloads, ensure the API is available (schedule new/replacement workloads), or jeopardize etcd-quorum.

Reply:
The MCO might need to be alive to run existing workloads.  For example, if your pull-secret needs rotating, or your Kube-API CA is getting rolled, etc., those are things that get pushed out to nodes via the machine-config operator.  And if that isn't getting pushed out, the nodes may die.  And possibly be destroyed.  And new nodes (replacements or autoscaling increases) will not come up without sufficient machine-config stack health.  So lots of vulnerabilities.  But vulnerabilities are not necessarily page-at-midnight until the lack of capacity or stale-pull-secret or other fallout is actually biting you.  And "core MCO operator is dead" shouldn't happen all that often anyway.  Is this really something where noise is a problem?  Or are you extrapolating from a different machine-config alert that you do find too noisy?

Reply:
It sounds like in this case there should be independent monitoring of this particular situation for kube-api CA's getting rolled.  So, the alert would be something along the lines of "CA not synced && MCO dead".

Reply:
"&& MCO dead" doesn't sound like something the CA-generator should be taught to care about.  But kubelets could provide metrics about remaining CA time, and there could be a new alert about those expiration times dropping under a given threshold?  I think it's fine to leave it to the user or local Insights-style rules to consolidate a collection of distinct alerts into diagnoses.  There is a distinction between "there is a problem" (alerts) and "I have a theory for the root cause" (diagnoses), and if we attempt to only target diagnoses with alerts, we risk silence when there is some new failure mode or presentation that doesn't match an existing diagnosis fingerprint.

Reply:
I think this is wrong example. MCO for sure is on the critical path and it should alert appropriately as critical component due to all reasons above and that it is actioner of a lot of control plane actions. And alerting in kube manager for "Cert rotation stuck" without giving me where to look is almost useless. But Alerting in kube manager with "Cert Rotation stuck. Rotation path: Manager -> MCO-> node" Gives me hint what to check and which alerts to look for next. So I dont see MCO alert in critical I know it works fine and go to the next in line. 
Every component and alert needs to understand their dependency flow in the the system overall.

`s/MCO/samples operator/` and your example should be good.

Reply:
I disagree.  If Cert rotation is the thing that will drag the cluster down, then we need to set an alert for that specifically.  We also aren't (shouldn't be?) waiting until the last possible minute to do cert rotation, we have plenty of time to take action on that before they expire, on the order of weeks IIRC.  So, if you have an MCO that goes down, it's not immediately critical.  Yes, if you completely neglect your cluster for weeks on end, it will fail, but the MCO itself going down is not going to prevent application communication.